src/core/net/dnssd_server.cpp

```diff
@@ -839,6 +855,121 @@ Error Server::ResolveByQueryCallbacks(Header                 &aResponseHeader,
     return error;
 }
 
+#if OPENTHREAD_CONFIG_DNS_UPSTREAM_QUERY_ENABLE
+bool Server::ShouldForwardToUpstream(const Header &aRequestHeader, const Message &aRequestMessage)
+{
+    bool     ret = true;
+    uint16_t readOffset;
+    char     name[Name::kMaxNameSize];
+
+    VerifyOrExit(aRequestHeader.IsRecursionDesiredFlagSet(), ret = false);
+    readOffset = sizeof(Header);
+
+    for (uint16_t i = 0; i < aRequestHeader.GetQuestionCount(); i++)
+    {
+        VerifyOrExit(kErrorNone == Name::ReadName(aRequestMessage, readOffset, name, sizeof(name)), ret = false);
+        readOffset += sizeof(Question);
+
+        VerifyOrExit(!Name::IsSubDomainOf(name, kDefaultDomainName), ret = false);
+        for (const char *blockedDomain : kBlockedDomain)
+        {
+            VerifyOrExit(!Name::IsSameDomain(name, blockedDomain), ret = false);
+        }
+    }
+
+exit:
+    return ret;
+}
+
+void Server::OnUpstreamQueryResponse(UpstreamQueryTransaction &aQueryTransaction, Message &aResponseMessage)
+{
+    Error err = kErrorNone;
+
+    VerifyOrExit(aQueryTransaction.IsValid());
+
+    err = mSocket.SendTo(aResponseMessage, aQueryTransaction.GetMessageInfo());
+    if (err != kErrorNone)
+    {
+        aResponseMessage.Free();
+    }
+
+    aQueryTransaction.Reset();
+    ResetTimer();
+
+exit:
+    return;
+}
+
+Server::UpstreamQueryTransaction *Server::AllocateUpstreamQueryTransaction(const Header           &aRequestHeader,
+                                                                           const Ip6::MessageInfo &aMessageInfo)
+{
+    UpstreamQueryTransaction *ret = nullptr;
+
+    for (UpstreamQueryTransaction &txn : mUpstreamQueryTransactions)
+    {
+        if (!txn.IsValid())
+        {
+            ret = &txn;
+            txn.Init(aRequestHeader, aMessageInfo);
+            LogInfo("Allocated new transaction for upstream query.");
```

Comment:
About logs, suggest thinking if/how they are useful when we debug issue on devices:

May be add more info here: query sender address, or ...? something to tell which txn it is?

Reply:
I moved some logs to the transaction object's Init and Reset.

Now, it prints:
```
[I] DnssdServer---: Upstream query transaction 0x557d79c75d88 initialized.                                                                                  
[I] Platform------: Forwarded DNS query 0x557d79c75d88 to 1 server(s) 
...
[I] Upstream query transaction 0x557d79c75d88 closed: ResponseTimeout
```